test/performance/odp_l2fwd.c

```diff
@@ -309,6 +320,211 @@ static inline int copy_packets(odp_packet_t *pkt_tbl, int pkts)
+				vpkt = odp_packet_vector_from_event(ev_tbl[i]);
+				pkts = odp_packet_vector_tbl(vpkt, &pkt_tbl);
+			} else  { /* ODP_EVENT_PACKET */
+				pkt  = odp_packet_from_event(ev_tbl[i]);
+				pkt_tbl = &pkt;
+				pkts = 1;
+			}
+
+			if (odp_unlikely(gbl_args->appl.extra_feat)) {
+				if (gbl_args->appl.packet_copy) {
+					int fails;
+
+					fails = copy_packets(pkt_tbl, pkts);
+					stats->s.copy_fails += fails;
+				}
+
+				if (gbl_args->appl.chksum)
+					chksum_insert(pkt_tbl, pkts);
+
+				if (gbl_args->appl.error_check) {
+					int rx_drops;
+
+					/* Drop packets with errors */
+					rx_drops = drop_err_pkts(pkt_tbl, pkts);
+
+					if (odp_unlikely(rx_drops)) {
+						stats->s.rx_drops += rx_drops;
+						if (pkts == rx_drops)
+							continue;
+
+						pkts -= rx_drops;
+						if (vpkt)
+							odp_packet_vector_size_set(vpkt, pkts);
+					}
+				}
+			}
+
+			/* packets from the same queue are from the same interface */
+			src_idx = odp_packet_input_index(pkt_tbl[0]);
+			ODPH_ASSERT(src_idx >= 0);
+			dst_idx = gbl_args->dst_port_from_idx[src_idx];
+			fill_eth_addrs(pkt_tbl, pkts, dst_idx);
+
+			if (odp_likely(use_event_queue)) {
+				odp_event_t event;
+
+				event = vpkt ?
+					odp_packet_vector_to_event(vpkt) : odp_packet_to_event(pkt);
+				sent = odp_queue_enq(tx_queue[dst_idx], event);
+				sent = odp_likely(sent == 0) ? pkts : 0;
+			} else {
+				sent = odp_pktout_send(pktout[dst_idx], pkt_tbl, pkts);
+				sent = odp_unlikely(sent < 0) ? 0 : sent;
+			}
+
+			tx_drops = pkts - sent;
+			if (odp_unlikely(tx_drops)) {
+				int j;
+
+				stats->s.tx_drops += tx_drops;
+				/* Drop rejected packets */
+				for (j = sent; j < pkts; j++)
+					odp_packet_free(pkt_tbl[j]);
+			}
+
+			/* Free packet vector if sending failed or in direct mode. */
+			if (vpkt && (tx_drops || !use_event_queue))
+				odp_packet_vector_free(vpkt);
+
+			stats->s.packets += pkts;
+		}
+	}
+
+	/*
+	 * Free prefetched packets before entering the thread barrier.
+	 * Such packets can block sending of later packets in other threads
+	 * that then would never enter the thread barrier and we would
+	 * end up in a dead-lock.
+	 */
+	odp_schedule_pause();
+	while (1) {
+		odp_event_t  ev;
+
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+		if (ev == ODP_EVENT_INVALID)
+			break;
+
+		if (odp_event_type(ev) == ODP_EVENT_PACKET_VECTOR)
+			free_packet_vector_tbl(odp_packet_vector_from_event(ev));
+
+		odp_event_free(ev);
+	}
+
+	/* Make sure that latest stat writes are visible to other threads */
+	odp_mb_full();
+
+	/* Wait until pktio devices are stopped */
+	odp_barrier_wait(&gbl_args->term_barrier);
+
+	/* Free remaining events in queues */
+	odp_schedule_resume();
+	while (1) {
+		odp_event_t  ev;
+
+		ev = odp_schedule(NULL,
+				  odp_schedule_wait_time(ODP_TIME_SEC_IN_NS));
+
+		if (ev == ODP_EVENT_INVALID)
+			break;
+
+		if (odp_event_type(ev) == ODP_EVENT_PACKET_VECTOR)
+			free_packet_vector_tbl(odp_packet_vector_from_event(ev));
```

Comment:
`odp_event_free()` frees also the possible packets inside a vector, so this code is unnecessary.

Reply:
Agree.